File: binarySearch/binarySearch.js
/*
basic binary search algorithm
return null if target is not found
*/

var binarySearch = (sortedArray, target) => {
  if (!Array.isArray(sortedArray)) {
    return 'ERROR: input is not an array';
  }
  if (sortedArray.length === 1) {
    if (sortedArray[0] === target) {
      return 0;
    }
  }
  var low = 0;
  var high = sortedArray.length - 1;

  while (low !== high) {
    var mid = Math.floor((low + high)/2);
    var guess = sortedArray[mid];
    if (guess === target) {
      return mid;
    }
    if (guess < target) {
      low = mid + 1;
    } else {
      high = mid - 1;
    }
  }
  return null;
};

export default binarySearch;

File: general/binarySearch/binarySearch.py
def binary_search(sorted_array, target) :
  low = 0
  high = len(sorted_array) - 1
  while low <= high:
    mid = (low + high) / 2
    guess = sorted_array[mid]
    if guess == target:
      return mid
    if guess > target:
      high = mid - 1
    else:
      low = mid + 1
  return None

my_list = [1, 3, 5, 7, 9]

print(binary_search(my_list, 3)) # should be 1
print(binary_search(my_list, -1)) # should be None